quinn-proto/src/bytes_source.rs

```diff
@@ -0,0 +1,231 @@
+//! Defines types which can be lazily converted into `Bytes` chunks
+
+use bytes::Bytes;
+
+/// A source of one or more buffers which can be converted into `Bytes` buffers on demand
+///
+/// The purpose of this data type is to defer conversion as long as possible,
+/// so that no heap allocation is required in case no data is writable.
+pub trait BytesSource {
+    // Returns the total amount of bytes to write
+    fn len(&self) -> usize;
+    /// Limits the source to the new size
+    fn limit(&mut self, new_len: usize);
```

Comment:
Rather than an extra method and associated state, could we add a `max_len: usize` to `pop_chunk`? We'd have to pass a limit from `Streams` to `streams::Send`, but that seems inoffsensive.

Reply:
It can be. But then we have to subtract the limit in `Stream::write` based on what was already written. I feel like then actually more code with less test coverage is tasked with keeping track of limits. This way its fairly concentrated in one place.

Reply:
It seems like the current approach has to duplicate the logic across both impls, whereas moving it out of the trait would yield a single implementation, and a trait that's simpler to reason about besides.